hnn_core/parallel_backends.py

```diff
@@ -0,0 +1,265 @@
+
+from warnings import warn
+
+_backend = None
+_loaded_dll = None
+
+
+class Joblib_backend(object):
+    """The Joblib_backend class.
+
+    Parameters
+    ----------
+    n_jobs : int
+        The number of jobs to start in parallel
+
+    Attributes
+    ----------
+    n_jobs : int
+        The number of jobs to start in parallel
+    """
+    def __init__(self, n_jobs=1):
+        self.type = 'joblib'
+
+        self.n_jobs = n_jobs
+        print("joblib will run over %d jobs" % (self.n_jobs))
+
+    def _parallel_func(self, func):
+        if self.n_jobs != 1:
+            try:
+                from joblib import Parallel, delayed
+            except ImportError:
+                warn('joblib not installed. Cannot run in parallel.')
+                self.n_jobs = 1
+        if self.n_jobs == 1:
+            my_func = func
+            parallel = list
+        else:
+            parallel = Parallel(self.n_jobs)
+            my_func = delayed(func)
+
+        return parallel, my_func
+
+    def __enter__(self):
+        global _backend
+        self._old_backend = _backend
+        _backend = self
+
+        return self
+
+    def __exit__(self, type, value, traceback):
+        global _backend
+        _backend = self._old_backend
+
+    def _clone_and_simulate(self, net, trial_idx):
+        # avoid relative lookups after being forked by joblib
+        from hnn_core.neuron import _neuron_network, _simulate_single_trial
+
+        if trial_idx != 0:
+            net.params['prng_*'] = trial_idx
+
+        neuron_net = _neuron_network(net.params)
+        dpl = _simulate_single_trial(neuron_net)
+
+        spikedata = neuron_net.get_data_from_neuron()
+
+        return dpl, spikedata
+
+    def simulate(self, net):
+        """Simulate the HNN model
+
+        Parameters
+        ----------
+        net : Network object
+            The Network object specifying how cells are
+            connected.
+
+        Returns
+        -------
+        dpl: list of Dipole
+            The Dipole results from each simulation trial
+        """
+
+        n_trials = net.params['N_trials']
+        dpls = []
+
+        parallel, myfunc = self._parallel_func(self._clone_and_simulate)
+        data = parallel(myfunc(net, idx) for idx in range(n_trials))
+
+        # the assignments below need to be made after any forking
+        for idx in range(n_trials):
+            dpls.append(data[idx][0])
+            spikedata = data[idx][1]
+            net.spiketimes.append(spikedata[0])
+            net.spikegids.append(spikedata[1])
+            net.gid_dict = spikedata[2]  # only have one gid_dict
+
+        return dpls
+
+
+class MPI_backend(object):
+    """The MPI_backend class.
+
+    Parameters
+    ----------
+    n_procs : int | None
+        The number of processors (cores) to start MPI job over
+    mpi_cmd : str
+        The name of the mpi launcher executable. Will use 'mpiexec'
+        (openmpi) by default.
```

Comment:
what are the other options? can we raise a ValueError if it's not one of the valid options? Or is it taken care of by mpi4py? It's better to raise errors early ...

Reply:
There's a huge list. I would prefer to build on this functionality as we need it. BTW, this is not `mpi4py`. We are calling the`mpiexec` binary. Some supercomputer clusters use `ibrun` or `srun` for example.